### DISCUSSION TITLE
Trying to run the MODE wrapper

### DISCUSSION THREAD

**Discussion by jcorner1** [dtcenter/METplus#1924] [Configuration] [answered]
Hello, 

I am trying to run the MODE wrapper with composite radar data as the obs and HRRR as the model input. I have tried a few different things but ultimately have failed get any results. I will send my data through the ftp server in the "corner_data" directory as well as the log file with the error as well the config file to run MODE.

Thanks,
Jeremy

**Comment by georgemccabe** [comment]
Hi @jcorner1, I pulled your files from FTP, but it looks like one of the files you provided may have been corrupted when it was uploaded. I am not able to read it using ncdump. Here is what I see:

> ncdump: Level3_Composite_dhr_1km_20221104_2100.gini.nc4: NetCDF: Unknown file format

Could you upload this file again?

**Reply by georgemccabe** [reply]
I'm also unable to open the grib2 file. Could you upload that again as well?
`wgrib2  hrrr.t17z.wrfsfcf04.grib2`

> *** FATAL ERROR: rd_grib2_msg_seq_file, missing end section ('7777') ***

**Reply by jcorner1** [reply]
Hi @georgemccabe, I have resent those files again. hopefully this fixes the issues. Thanks!

**Reply by georgemccabe** [reply]
Hi @jcorner1, I downloaded the new files but I am still getting the same errors. How are these files being created? Is there a way I can download these files directly? Are you able to run 'ncdump' on the nc4 file or wgrib2 on the grib2 file to view the contents? I have seen the 7777 grib2 error before when trying to process a file that was still being written to disk from the source.

**Reply by jcorner1** [reply]
Hi @georgemccabe, the HRRR file can be download from here (https://pando-rgw01.chpc.utah.edu/hrrr/sfc/20221104/hrrr.t17z.wrfsfcf04.grib2) and the composite radar file from (https://thredds.ucar.edu/thredds/catalog/nexrad/composite/gini/dhr/1km/20221104/catalog.html?dataset=Unidata/NEXRAD/NationalComposite/GINI/dhr/1km/20221104/Level3_Composite_dhr_1km_20221104_2100.gini) . I did still have the file open and parsing the data when I sent the file, so I believe that was the issue. If there is a problem I will apologize for miraculously corrupting the file on the server!

**Reply by georgemccabe** [reply]
Hi @jcorner1, thanks. I was able to download the grib2 file and it opens without issue in wgrib2. It looks like the radar file is the raw GINI input and not the NetCDF file that you are using. Did you convert the file yourself?

Having the file open when you uploaded it may have caused the issue. Could you upload the nc4 file again after you have closed the file?

**Reply by jcorner1** [reply]
Hi @georgemccabe, That might highlight an issue I am having as I have not converted out of the .gini file to a straight .nc4 file yet. Is there a preferred method of converting the file. Running the ncdump command does produce the radar reflectivity values.

**Reply by georgemccabe** [reply]
Ah, I thought that since the files you sent had the .nc4 extension that you had converted them to NetCDF. MET does not read GINI data natively. I believe there are utilities to convert them to NetCDF, but I am personally not familiar with any. @JohnHalleyGotway may have a suggestion.

If there exists a Python library that can read GINI data, then you could utilize Python Embedding to read in the data directly into the MET tools.

**Reply by georgemccabe** [reply]
I looked closer at the THREDDS page for the GINI data, and it looks like you can covert it to NetCDF through THREDDS. I just did this and was able to download a nc4 file that I can open with ncdump. I am out on PTO tomorrow but I can take a closer look at this use case on Monday.

**Reply by jcorner1** [reply]
Sounds good to me, I'm in no current time crunches but in the mean time I will figure out how to convert the file through THREDDS to a nc file. Enjoy time off and thanks for the help!

**Reply by jcorner1** [reply]
Hi @georgemccabe,

Sorry for the delay. Have been working with some new datasets as well as doing some regridding, so its been a bit since I have worked with MODE. Have gotten past the issue with metplus reading the datafiles and now have a problem with the gridding of the data. I have regridded both datasets to a HRRR file from one of the MODE examples. However it still says the lons are causing problems in the log. I will send all the files in through the NCAR site in the 'corner_data' directory.

**Reply by georgemccabe** [reply]
Hi @jcorner1, I was out of the office for over a week so I am just now getting back to this. I am unable to see any files in corner_data. They may have been scrubbed. If the input files are the same as what you previously provided, you could try attaching the config file to this discussion. You may need to rename the file to include a .txt extension.

**Reply by jcorner1** [reply]
Hi @georgemccabe, I have sent my data back through in my directory. I have changed how the data is done from the first post. Thanks!

**Reply by georgemccabe** [reply]
Hi @jcorner1, unfortunately it looks like those data files are corrupted again. Is there another way you could provide those files?

I was able to look at the configuration file and log output. MET does not support lat/lon grids that are not evenly spaced. You could use a [Python Embedding](https://met.readthedocs.io/en/latest/Users_Guide/appendixF.html#python-embedding-for-2d-data) script to preprocess the data to create an evenly spaced grid. If you have other tools available to regrid these data, the commands to run those tools can be incorporated into your METplus use case by using the [UserScript ](https://metplus.readthedocs.io/en/latest/Users_Guide/wrappers.html#userscript) wrapper. You may be able to prevent the error by using the FORCE regrid option, but that isn't the best solution because it would produce results that are slightly incorrect.

**Reply by jcorner1** [reply]
Hi @georgemccabe, I have a bit of confusion of because of this error due to me having regridded the data to an example before. That's why I'm confused I am getting this error because I do not feel that I should be if the example is working. I'm using xesmf and it seems to be working fine otherwise. I will try to send over my data again. I'm unsure of why it keeps becoming corrupt. This time the data is being created by me writing it to the new script via the information online. Is it possible this is where the mistake lies and that I don't have CF conventions correct or something like that? or is it possible the lat-lons are hiding in the file where MET can't get to?

**Reply by georgemccabe** [reply]
I'm not sure I follow. I don't see any regridding options set in the config and log that you sent. Are you saying that you regridded the data prior to this use case and are passing in the output from that run to this case? If so, what did you use to regrid the data? If you have regridded the data prior to running METplus and are confident that the grids match, then you could try using `MODE_REGRID_METHOD = FORCE`.

The HRRR file that you uploaded to FTP is much smaller than the file I was able to download from the link you provided, so I'm guessing the upload is being cancelled before it can complete. Maybe there are firewall restrictions preventing the upload. Is there another way you can make these data available to us?

**Reply by jcorner1** [reply]
Hi @georgemccabe, That is correct, I regridded the file prior to running METplus. I would think the HRRR file is smaller as I eliminated everything in the file that was unneeded to save space, so it should be just the datetime, lats/lons, and reflectivity. 

I did try setting the `MODE_REGRID_METHOD = FORCE` to the config script, but I still got the same error. I'm using "xesmf" to do the regridding and using a hrrr_esrl file to regrid to so that I know the grid is correct as it works in one of the example cases, however I am still getting the error with the lat/lons not being evenly spaced. That is my point of confusion on why the example works but when I do it I get the lat/lon error.

**Reply by georgemccabe** [reply]
It looks like the hrrr_esrl data from the short_range/MODE_fcstHRRR_obsMRMS_Hail_GRIB2 use case is on a Lambert Conformal grid, so I'm confused as to why the data you regridded is on a lat/lon grid. I'm not sure how xesmf regrids data but this may be causing an issue, as NetCDF files that have unevenly lat/lon grids cannot be read natively by MET.

I noticed that the example use case you are referring to reads in GRIB2 observations and regrids the data to the forecast grid in MODE using `MODE_REGRID_TO_GRID = FCST`. If the MRMS data is GRIB2, you could try reading it into MODE without preprocessing it and let MET handle the regridding.

Is there another way you can provide us with the input data? It would be easier to see what exactly is going on with access to those files. You could upload them to a Google Drive folder or perhaps try uploading them to the FTP from another computer.

**Reply by jcorner1** [reply]
Hi @georgemccabe. 
I'm not sure what the issue is with the file but I have uploaded them to a Google Drive (https://drive.google.com/drive/folders/1BwWrdUUEcG0xRc8IU0BzZ_hYUxcUC0ib?usp=sharing). If it doesn't work, I will assume that it is an issue with how am creating these data or how it is regridded by "xesmf" (possibly forcing it to change projection also). I wasn't able to work on your other ideas of fixes yesterday, but today I will give them a try as well as look at other test cases to see if we can come to a solution. Sorry for all the trouble with the data thus far, hopefully Google Drive will fix this issue. Thanks!

**Reply by georgemccabe** [reply]
Hi @jcorner1, I was able to download the files from the Google Drive folder and they are not corrupted. Hooray! It looks like both of the files are NetCDF despite the grib2 extension of the HRRR file. It looks like the HRRR file is lacking metadata required to read it into MET. The MRMS file also cannot be read due to the unevenly spaced lat/lon grids that we discovered earlier.

Could you upload the input files that were used to generate these files and the script you used to regrid them?

**Reply by jcorner1** [reply]
Hi @georgemccabe 
I should have put them in the Google Drive now. let me know if you need anything else!

I did catch the thing with the extension being the wrong format and hopefully that will clear any futures that will cause. I did try it again, but obviously that didn't fix the lat/lon issue. 

NOTE:
the MRMS data is the "res.nc" file. That was the name the file was given when I downloaded from the sever manually and I didn't catch that until after I uploaded it to the Google Drive.

**Comment by georgemccabe** [comment, marked as answer]
Thanks for sending those files. It looks like I am able to read both the HRRR and the MRMS files into MET.

![hrrr_mrms_met_images](https://user-images.githubusercontent.com/23407799/206002800-0d80e833-ecf3-4454-b65e-41f366d2e86d.PNG)

Here are the commands I ran to generate the images:

```
plot_data_plane ./res.nc out_res.ps 'name="mrms_lcref"; level="(*,*)";'
convert -rotate 90 out_res.ps out_res.png
plot_data_plane ./hrrr.t16z.wrfsfcf05.grib2 out_hrrr.ps 'name="REFC"; level="L0";'
convert -rotate 90 out_hrrr.ps out_hrrr.png
```

You should be able to use the same name and level information to read the fields into MODE. Just be sure to set `MODE_REGRID_TO_GRID = FCST` or `MODE_REGRID_TO_GRID = OBS` so that the data are matched to the same grid.

Is there a reason you need to regrid these data with a script before passing them to MET? It looks like that step may not be necessary.

**Reply by jcorner1** [reply]
Hi @georgemccabe,

My advisor wants it regridded but I don't have a specific reason at the current moment. We'll be doing some machine learning, so I'm guessing that's his reasoning for now. I don't think it's an issue right now.

I'm having a little problem with setting the level in the .conf file. I will attach my log and conf file as well as put them in the Google Drive as I no longer trust a single way of sending files. 
[MODE_HRRR_FCST_RADAR_OBS.conf.txt](https://github.com/dtcenter/METplus/files/10170275/MODE_HRRR_FCST_RADAR_OBS.conf.txt)
[metplus.log.20221206145041.txt](https://github.com/dtcenter/METplus/files/10170276/metplus.log.20221206145041.txt)

**Reply by georgemccabe** [reply]
Hi @jcorner1, you can set `MODE_REGRID_TO_GRID` to the path to a file that MET can read and it will use its grid. You can also define an explicit grid definition. See [Grid Definition Strings](https://met.readthedocs.io/en/latest/Users_Guide/appendixB.html#grid-specification-strings) in the MET User's Guide.

It looks like you have `refc` set for the forecast field name, but the name of the field is `REFC`.

**Comment by jcorner1** [comment]
That looks to have done it! Thanks @georgemccabe!!!

**Reply by georgemccabe** [reply]
Great to hear! I am going to lock this discussion to prevent future posts. If you have any other questions, please create a new discussion topic.
